### DISCUSSION TITLE
Is it possible to extract the jacobian matrix of a system with SofaPython ?

### DISCUSSION THREAD

**Discussion by Matyas-D** [sofa-framework/sofa#3350] [Develop in C++] [answered]
Hi everyone,

I am currently working on a parallel continuum robot modeled with SOFA. I know the jacobian matrix of the system/robot is computed somewhere whithin SOFA, and I need it for my control loop. Is there any way to extract this matrix ? 

This discussion: [https://github.com/sofa-framework/sofa/discussions/2912](url) shows that there was an ongoing work on making system matrices accessible, is it ready now ?

Thank you for your help.

**Comment by alxbilger** [comment, marked as answer]
Hi @Matyas-D

if you use Python, you can have read-access to various matrices and vectors in the simulation scenes:
- Global system matrix
- Global system right-hand side
- Global system solution
- Mass matrix
- Stiffness matrix
- Constraint matrix

Dense objects will be available as a numpy array, while sparse matrices will be represented by scipy.sparse.csr_matrix.

This is available in SOFA v22.06.

Checkout the examples in the SofaPython3 plugin. For example:

https://github.com/sofa-framework/SofaPython3/blob/fffebadfbe4e68a1adf1910efa622e27af59209c/examples/access_matrix.py#L44

Alex

**Reply by Matyas-D** [reply]
Hi @alxbilger,

Thank you for your fast answer. It is indeed what I am looking for, unfortunately I have already looked into the solvers but I don't have access to these elements even with SOFA v22.06.99. If I simply use the provided code with no modification I get the following error:

> [ERROR]   [MatrixAccessController(matrixAccessor)] Unable to execute code.
> Python exception:
>   KeyError: 'A'

And so on with the others matrices and vectors. Maybe it is only a bug comming from my SOFA installation.

Matyas

**Reply by alxbilger** [reply]
Are you trying to execute the example `access_matrix.py` ? If not, try the example to see if it works.
What version of SofaPython3 are you running?
Are you compiling from sources or are you using the binaries?

**Reply by Matyas-D** [reply]
Yes I was trying to execute the example `access_matrix.py` as is, whith the version SOFA v22.06.99, python 3.8.10. I am using the binaries provided at the SOFA summer school.

**Reply by alxbilger** [reply]
Could you try with the official binaries? https://www.sofa-framework.org/download/

**Reply by Matyas-D** [reply]
Good news, this worked ! The example runs smoothly as well as my own code with the official binaries. Thank you for your help Alex. 
I will mark your first answer as solution as it is the more relevant for future readers.
